docs/doxygen/man/man3/ChangeNodeKey.3
.TH "ChangeNodeKey" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ChangeNodeKey \- Class responsible for handling 'change_node_key' command\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <changenodekey\&.h>\fP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBChangeNodeKey\fP (uint16_t srcBank, uint32_t srcUser, uint32_t msid, uint16_t dstNode, uint32_t time, uint8_t pubkey[32])"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_BKY as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eModifying as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet user message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetDestBankId\fP ()"
.br
.RI "\fIGet destination bank id\&. \fP"
.ti -1c
.RI "virtual uint8_t * \fBgetKey\fP ()"
.br
.RI "\fIGet new public key\&. \fP"
.ti -1c
.RI "virtual void \fBsetOldPublicKey\fP (uint8_t *data)"
.br
.RI "\fISet old public key\&. \fP"
.ti -1c
.RI "virtual uint8_t * \fBgetOldPublicKey\fP ()"
.br
.RI "\fIGet old public key\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetBlockMessage\fP () override"
.br
.RI "\fIGet data as it is stored in blockchain\&. \fP"
.ti -1c
.RI "virtual size_t \fBgetBlockMessageSize\fP () override"
.br
.RI "\fIGet blockchain data size\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.RI "\fIGet additional data size\&. \fP"
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Class responsible for handling 'change_node_key' command\&. 
.PP
Definition at line 11 of file changenodekey\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool ChangeNodeKey::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 63 of file changenodekey\&.cpp\&.
.SS "bool ChangeNodeKey::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 106 of file changenodekey\&.cpp\&.
.SS "void ChangeNodeKey::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 58 of file changenodekey\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/changenodekey.cpp
#include "changenodekey.h"
#include <iostream>
#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/hash.h"
#include "helper/json.h"

ChangeNodeKey::ChangeNodeKey()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

ChangeNodeKey::ChangeNodeKey(uint16_t srcBank, uint32_t srcUser, uint32_t msid, uint16_t dstNode, uint32_t time, uint8_t pubkey[32])
    : m_data(srcBank, srcUser, msid, time, dstNode, pubkey) {
    m_responseError = ErrorCodes::Code::eNone;
}

unsigned char* ChangeNodeKey::getData() {
    return reinterpret_cast<unsigned char*>(&m_data);
}

unsigned char* ChangeNodeKey::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void ChangeNodeKey::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void ChangeNodeKey::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int ChangeNodeKey::getDataSize() {
    return sizeof(m_data) - getSignatureSize() - sizeof(m_data.old_public_key);
}

int ChangeNodeKey::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* ChangeNodeKey::getSignature() {
    return m_data.sign;
}

int ChangeNodeKey::getSignatureSize() {
    return sizeof(m_data.sign);
}

int ChangeNodeKey::getType() {
    return TXSTYPE_BKY;
}

CommandType ChangeNodeKey::getCommandType() {
    return CommandType::eModifying;
}

void ChangeNodeKey::sign(const uint8_t* hash, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign2(hash, SHA256_DIGEST_LENGTH, getData(), getDataSize(), sk, pk, getSignature());

}

bool ChangeNodeKey::checkSignature(const uint8_t* hash, const uint8_t* pk) {
    return( ed25519_sign_open2(hash, SHA256_DIGEST_LENGTH, getData(), getDataSize(), pk, getSignature() ) == 0);
}

uint32_t ChangeNodeKey::getUserId() {
    return m_data.src_user;
}

uint32_t ChangeNodeKey::getBankId() {
    return m_data.src_node;
}

uint32_t ChangeNodeKey::getTime() {
    return m_data.ttime;
}

unsigned char*  ChangeNodeKey::getBlockMessage()
{
    return reinterpret_cast<unsigned char*>(&m_data);
}

size_t  ChangeNodeKey::getBlockMessageSize()
{
    return sizeof(ChangeNodeKeyData);
}

/*unsigned char*  ChangeNodeKey::getAdditionalData()
{
    return reinterpret_cast<unsigned char*>(&m_data.old_public_key);
}

int ChangeNodeKey::getAdditionalDataSize() {
    return sizeof(m_data.old_public_key);
}*/

int64_t ChangeNodeKey::getFee() {
    return TXS_BKY_FEE;
}

int64_t ChangeNodeKey::getDeduct() {
    return 0;
}

bool ChangeNodeKey::send(INetworkClient& netClient)
{
    if(! netClient.sendData(getData(), sizeof(m_data) )) {
        ELOG("ChangeNodeKey sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("ChangeNodeKey reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if(!netClient.readData(getResponse(), getResponseSize())) {
        ELOG("ChangeNodeKey ERROR reading global info\n");
        return false;
    }

    return true;
}

void ChangeNodeKey::saveResponse(settings& sts) {
    if (!sts.without_secret && !std::equal(sts.pk, sts.pk + SHA256_DIGEST_LENGTH, m_response.usera.pkey)) {
        m_responseError = ErrorCodes::Code::ePkeyDiffers;
    }

    std::array<uint8_t, SHA256_DIGEST_LENGTH> hashout;
    Helper::create256signhash(getSignature(), getSignatureSize(), sts.ha, hashout);
    if (!sts.signature_provided && !std::equal(hashout.begin(), hashout.end(), m_response.usera.hash)) {
        m_responseError = ErrorCodes::Code::eHashMismatch;
    }

    sts.msid = m_response.usera.msid;
    std::copy(m_response.usera.hash, m_response.usera.hash + SHA256_DIGEST_LENGTH, sts.ha.data());
}

uint32_t ChangeNodeKey::getDestBankId() {
    return m_data.dst_node;
}

uint8_t* ChangeNodeKey::getKey() {
    return m_data.node_new_key;
}

void ChangeNodeKey::setOldPublicKey(uint8_t *data) {
    std::copy(data, data + 32, m_data.old_public_key);
}

uint8_t* ChangeNodeKey::getOldPublicKey() {
    return m_data.old_public_key;
}

uint32_t ChangeNodeKey::getUserMessageId() {
    return m_data.msg_id;
}

std::string ChangeNodeKey::toString(bool /*pretty*/) {
    return "";
}

void ChangeNodeKey::toJson(boost::property_tree::ptree& ptree) {
    if (!m_responseError) {
        ptree.put("result", "Node key changed");
        Helper::print_user(m_response.usera, ptree, true, this->getBankId(), this->getUserId());
        Helper::print_msgid_info(ptree, m_data.src_node, m_response.msid, m_response.mpos);
    } else {
        if (m_responseError == ErrorCodes::Code::ePkeyDiffers) {
            std::stringstream tx_user_hashin{};
            Helper::ed25519_key2text(tx_user_hashin, m_response.usera.pkey, SHA256_DIGEST_LENGTH);
            ptree.put("tx.account_public_key_new", tx_user_hashin.str());
        }
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void ChangeNodeKey::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.ttype));
    ptree.put(TAG::SRC_NODE, m_data.src_node);
    ptree.put(TAG::SRC_USER, m_data.src_user);
    ptree.put(TAG::MSGID, m_data.msg_id);
    ptree.put(TAG::TIME, m_data.ttime);
    ptree.put(TAG::DST_NODE, m_data.dst_node);
    ptree.put(TAG::OLD_PKEY, ed25519_key2text(m_data.old_public_key, sizeof(m_data.old_public_key)));
    ptree.put(TAG::NEW_PKEY, ed25519_key2text(m_data.node_new_key, sizeof(m_data.node_new_key)));
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string ChangeNodeKey::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"change_node_key\",\"public_key\":<public_key>,[\"node\":<node_id>]}" << "\n";
    ss << "Example: " << "{\"run\":\"change_node_key\",\"public_key\":\"2D1FC97FA56B785E0FDAE5752DE613BAD7FBBB5EBBB46DAEE5DBFA822F976B63\",\"node\":\"2\"}" << "\n";
    return ss.str();
}
